java/ql/src/experimental/Security/CWE/CWE-522/UnsecureBasicAuth.ql

```diff
@@ -0,0 +1,228 @@
+ */
+class TypeHttpUrlConnection extends RefType {
+  TypeHttpUrlConnection() { hasQualifiedName("java.net", "HttpURLConnection") }
+}
+
+/**
+ * String of HTTP URLs not in private domains.
+ */
+class HttpString extends StringLiteral {
+  HttpString() {
+    // Match URLs with the HTTP protocol and without private IP addresses to reduce false positives.
+    exists(string s | this.getRepresentedString() = s |
+      s = "http"
+      or
+      s.matches("http://%") and
+      not s.matches("%/localhost%") and
+      not s.matches("%/127.0.0.1%") and
+      not s.matches("%/10.%") and
+      not s.matches("%/172.16.%") and
+      not s.matches("%/192.168.%")
+    )
+  }
+}
+
+/**
+ * String concatenated with `HttpString`.
+ */
+predicate builtFromHttpStringConcat(Expr expr) {
+  expr instanceof HttpString
+  or
+  builtFromHttpStringConcat(expr.(AddExpr).getLeftOperand())
+  or
+  exists(Expr other | builtFromHttpStringConcat(other) |
+    exists(Variable var | var.getAnAssignedValue() = other and var.getAnAccess() = expr)
+  )
+  or
+  exists(Expr other | builtFromHttpStringConcat(other) |
+    exists(ConstructorCall cc |
+      (
+        cc.getConstructedType().hasQualifiedName("org.apache.http.message", "BasicRequestLine") and
+        cc.getArgument(1) = other // BasicRequestLine(String method, String uri, ProtocolVersion version)
+        or
+        cc.getConstructedType().hasQualifiedName("java.net", "URI") and
+        cc.getArgument(0) = other // URI(String str) or URI(String scheme, ...)
+      ) and
+      (
+        cc = expr
+        or
+        exists(Variable var, VariableAssign va |
+          var.getAnAccess() = expr and
+          va.getDestVar() = var and
+          va.getSource() = cc
+        )
+      )
+    )
+  )
+}
+
+/**
+ * String pattern of basic authentication.
+ */
+class BasicAuthString extends StringLiteral {
+  BasicAuthString() { exists(string s | this.getRepresentedString() = s | s.matches("Basic %")) }
+}
+
+/**
+ * String concatenated with `BasicAuthString`.
+ */
+predicate builtFromBasicAuthStringConcat(Expr expr) {
+  expr instanceof BasicAuthString
+  or
+  builtFromBasicAuthStringConcat(expr.(AddExpr).getLeftOperand())
+  or
+  exists(Expr other | builtFromBasicAuthStringConcat(other) |
+    exists(Variable var | var.getAnAssignedValue() = other and var.getAnAccess() = expr)
+  )
+}
+
+/**
+ * The methods `addHeader()` and `setHeader` declared in ApacheHttpRequest invoked for basic authentication.
+ */
+class AddBasicAuthHeaderMethodAccess extends MethodAccess {
+  AddBasicAuthHeaderMethodAccess() {
+    this.getMethod().getDeclaringType() instanceof ApacheHttpRequest and
+    (this.getMethod().hasName("addHeader") or this.getMethod().hasName("setHeader")) and
+    this.getArgument(0).(CompileTimeConstantExpr).getStringValue() = "Authorization" and
+    builtFromBasicAuthStringConcat(this.getArgument(1)) and
+    exists(VarAccess request, VariableAssign va, ConstructorCall cc, Expr arg0 |
```

Comment:
For `SetBasicAuthPropertyMethodAccess` you used `Dataflow`, but here you only check for a direct variable access. I think using `Dataflow` for both would be more general.

Reply:
Thanks @aibaars for the advice. I've made the query more general.